### PR TITLE
verify debug usage: check for parameterless toString() calls

### PR DESCRIPTION
**Pull request by davleopo**
cc @zapster

### CODE REVIEW

**Comment by zapster** [comment]
Looks good. Btw, I noticed that we are only checking `Debug.log`. We should do the same for `Debug.logAndIndent` but I think this is out of the scope of this PR.

**Comment by davleopo** [comment]
@zapster , yes maybe we should start checking all the other debug methods (not only the `log` calls) for invalid calls to `toString`. In think `Debug.log*`,`Debug.logAndIndent*` and also `Debug.dump*` are potential candidates for calls to `toString`

@dougxc what do you think?

File: graal/com.oracle.graal.phases/src/com/oracle/graal/phases/verify/VerifyDebugUsage.java [hunk 1]
```diff
@@ -60,8 +60,15 @@ protected boolean verify(StructuredGraph graph, PhaseContext context) {
...
-                                                        + "    Use one of the multi-parameter Debug.log() methods or Debug.logv() instead.", e, argIdx, callee.format("%H.%n(%p)")));
+                                        throw new VerificationError(String.format("%s: parameter %d of call to %s appears to be a String concatenation expression.%n" +
+                                                        "    Use one of the multi-parameter Debug.log() methods or Debug.logv() instead.", e, argIdx, callee.format("%H.%n(%p)")));
+                                    }
+                                    if (m.getSignature().getParameterCount(false) == 0 && m.getSignature().getReturnType(m.getDeclaringClass()).equals(
+                                                    context.getMetaAccess().lookupJavaType(String.class))) {
+                                        StackTraceElement e = graph.method().asStackTraceElement(invoke.bci());
+                                        throw new VerificationError(String.format("%s: parameter %d of call to %s appears to be a toString call without a parameter. %n " +
```
Comment (dougxc) [inline graal/com.oracle.graal.phases/src/com/oracle/graal/phases/verify/VerifyDebugUsage.java]: The error message could be clearer. Something like:
`"%s: parameter %d of call to %s is a call to toString() which is redundant (the callee will do it) and forces unnecessary eager evaluation"`

**Comment by dougxc** [comment]
Yes, I think we should broaden this checking to all relevant methods in `Debug`.

File: graal/com.oracle.graal.phases/src/com/oracle/graal/phases/verify/VerifyDebugUsage.java (hunk 1 above)
Comment (davleopo) [inline graal/com.oracle.graal.phases/src/com/oracle/graal/phases/verify/VerifyDebugUsage.java]: thanks, will do

**Comment by davleopo** [comment]
ok, then I will prepare a new PR inlcuding all checks for the relevant methods in `Debug`.
Should I still update this PR or can we close it and open a new one including all changes?

**Pull request merged**